### ISSUE TITLE
Incorrect output using Barracuda compared to torch when excuting a 1D vector input outputing a 2D image

### ISSUE THREAD

**Issue by senbinyu** [Unity-Technologies/barracuda-release#109]
Hi, 

Our team are using a network transfering a 1d meaningful vector as input (1, 1, 1, 90),  to a 2d image as output (1, 128, 128, 1). We trained our model well and got satisfactory results in the pytorch platform with both .pth and .onnx models. 

First of all, we believed that our implementation of the .onnx model into the Unity/Barracuda is correct, since we tested with 1D vector input to 1D vector output and 2D Mnist both successfully. The results from Barracuda closely match with that from pytorch platform. 

The incorrect outputs occurred when we are using 1d vector as the input.  After several FC layers, we reshaped the extended 1d vector into H*W*C, a 3d matrix, then the results are messed up. The figures from barracuda and torch are quite different.

![barracuda_pred](https://user-images.githubusercontent.com/42667259/99036851-6cd6df80-25bd-11eb-83d0-803c174cbc81.png)
![torch_pred](https://user-images.githubusercontent.com/42667259/99036856-6f393980-25bd-11eb-9945-2696ed839ed1.png)

Here I attached the .onnx model and one sample of input in a .cs script and output ("ground_truth.png" and "torch_pred.png") in a zip file. Thanks for your kind attention.

[barracuda_issues.zip](https://github.com/Unity-Technologies/barracuda-release/files/5535139/barracuda_issues.zip)

**Comment by AlexRibard** [comment]
thanks for the info. I will check

**Comment by senbinyu** [comment]
Hi, Alex, 

Thanks for your reply. Now we have located where this error occurred. 

We believed that it is caused by the re-allocate of the memory when the is is reshaped into multiply channels(if it is reshaped into W * H * 1 channel, there's no error). We trained a NN using 1D input and output a small image. By comparing to the torch output, we tested it with Barracuda first using [1,1,1,64] and reshaped into [1,8,8,1], everying is fine. But when we used [1, 1, 1, 192] as input and reshaped it into [1, 8, 8, 3], the sequence of vector is changed. Normally, the memory storage should be read out as [RGB, RGB,...] in image, but it seems that it is read out as [RRR...., GGGG..., BBB...] in Barracuda. We think this is a similar problem with that of issue #105 . Hope this can speed up the debuging.

Looking forward to your validation and debuging. Good luck.

**Comment by AlexRibard** [comment]
Hey @senbinyu which version are you on?
I tested your model against `onnx-runtime` and `Barracuda` on random inputs.
At first glace there is some slight precision discrepancies compared to onnx, but at ~10e-3 absolute error it's ok.
(not sure if this is acceptable given you are trying to output  as mask...)
But things looks ok considering inference logic.

I will look more in details at your C# project tomorrow. 
But at first glance I noticed that the input tensor shape is not constructed correctly 
```
        TensorShape tensorShape = new TensorShape(1, 1, rawInputs.Count, 1); // input is 1x90 -> initialize TensorShape(1,rawInputs.Count) instead
        Tensor input = new Tensor(tensorShape, rawInputs.ToArray());
```
Let me know if that helps,
I'll you know once I have more info

**Comment by senbinyu** [comment]
Hi, @AlexRibard,

Thanks a lot for your reply. I used Barracuda 1.0.4 before and now I updated it into 1.2.0 preview version. After changing the tensorShape as you mentioned before, there is still nothing changed for me.

**Comment by AlexRibard** [comment]
@senbinyu ok figured it out.
* I dumped the output tensor and displayed it in python
![image](https://user-images.githubusercontent.com/12990800/99292491-46a87e00-280f-11eb-95a7-4e6e70f71cd2.png)
So thing are correct, when comparing the tensors.
* back to your demo
The problem you are having is due to how the RenderTexture are interpreted in Unity vs python
https://docs.unity3d.com/Manual/SL-PlatformDifferences.html
So your coordinate is probably the opposite of Unity's.
Internally we flip the y axis to be consistent to Unity axis, but sadly do not currently haven't exposed that.
So you'll need to change it manually 
```
BarracudaReferenceCompute.cs
TensorToRenderTexture
        fn.shader.SetBool("_FlipY", true); // set that to  fn.shader.SetBool("_FlipY", false);
```
Or change the axis when training.
Here is the result without the flipped axis.
![image](https://user-images.githubusercontent.com/12990800/99293233-5bd1dc80-2810-11eb-89dc-f518f28223db.png)

Alternatively you can transpose the output tensor before calling `ToRenderTexture`
Like so
```
        X = Reshape(X, new TensorShape(128, 128));
        X = Transpose(X);
        X = Reshape(X, new TensorShape(1,128, 128,1));
```
![image](https://user-images.githubusercontent.com/12990800/99294716-71e09c80-2812-11eb-8566-72b6c35661ed.png)

**Comment by senbinyu** [comment]
Hi, @AlexRibard ,

Thanks a lot for your quick and kind reply. I guess I encountered some problems in Unity C# sript, since I am new about it. Can you share your C# sript with me? I can compare them and find out exactly where the problem is.
Thank you again and best wishes.

**Comment by AlexRibard** [comment]
Sure. My script is very similar to yours
[barracuda_issue_test.zip](https://github.com/Unity-Technologies/barracuda-release/files/5554656/barracuda_issue_test.zip)
The "fixes" I was mentioning were done on the file `BarracudaReferenceCompute.cs`
```
TensorToRenderTexture
either change
fn.shader.SetBool("_FlipY", true); // set that to  fn.shader.SetBool("_FlipY", false);

or add 
        X = Reshape(X, new TensorShape(X.height, X.width));
        X = Transpose(X);
        X = Reshape(X, new TensorShape(1,X.height, X.width,1));
before declaring fn
```
If you want to save the Tensor to file, you can use `float[] array = output.ToReadOnlyArray()`

**Comment by senbinyu** [comment]
Hi, @AlexRibard 

Thanks a lot for your kind reply. We tried both methods you provided, but the result is still inconsistent with what you obtained above. In fact, change the "_FlipY" true or false only affect the direction or transpose, not the output values. For us, the output tensor values are totally different from what you obtained and what we obtained from onnxruntime. We used Barracuda 1.0.4 and 1.2.0. 
I want to check with you which version of Barracuda are you using? We are wondering that if the problem for us may happen in the Excute() function. 

In detail, when set
`fn.shader.SetBool("_FlipY", true);`
![image](https://user-images.githubusercontent.com/42667259/99495177-362e0a00-29ad-11eb-8ff4-e8e8b85f430f.png)
and set 
`fn.shader.SetBool("_FlipY", false);`
![68080C79-5063-47f6-A769-271F20A3FC18](https://user-images.githubusercontent.com/42667259/99609422-d71ed280-2a4a-11eb-9fa0-d071062b7e04.png)

**Comment by AlexRibard** [comment]
ah! you are very correct. I only tested on latest.
It is indeed broken on 1.2.0. On this version I do get the same issue as you.
I'll figure out what got fixed and share it with you

**Comment by AlexRibard** [comment]
Hey sorry about the false lead, there was indeed some fixes needed in the importer in order to make your model work.
```
ONNXModelConverter.cs
* comment l180: 
//&& onnxShape[onnxShape.Length - 1] == -1
* change l200
var features = onnxShape.Length > 1 ? onnxShape[1] : -1;
Output(node, features: features, rank: onnxShape.Length);
```
With this change on 1.2.0 I get this result
![image](https://user-images.githubusercontent.com/12990800/99546830-d58ad700-2984-11eb-9706-282be32281e5.png)
which is correct (against onnx-runtime it's good at 10e-3), modulus the image orientation.

**Comment by senbinyu** [comment]
Hi, @AlexRibard 

Thanks for your great help. Now we can obtain the same results like yours.
